src/compiler/parser.ts

```diff
@@ -2247,6 +2247,13 @@ namespace ts {
                 property.name = name;
                 property.questionToken = questionToken;
                 property.type = parseTypeAnnotation();
+
+                // Although object type properties cannot not have initializers, we attempt to parse an initializer
+                // so we can report that an object type property cannot have an initializer.
+                if (token === SyntaxKind.EqualsToken && lookAhead(() => parseNonParameterInitializer()) !== undefined) {

```

Comment:
So to summarize @delta-nry, just add an optional initializer onto each PropertySignature and perform a grammar check in `checker.ts` on each property. Report the error on the initializer itself.